### ISSUE TITLE
Route generation is adding a / at the end of routes (like $project/$tab/index.tsx)

### ISSUE THREAD

**Issue by divmgl** [TanStack/router#1517]
### Describe the bug

Say I have this folder structure:

<img width="202" alt="Screenshot 2024-04-24 at 2 55 31 PM" src="https://github.com/TanStack/router/assets/5452298/549ea738-6b32-40e2-8415-0d47b3d9ac1d">

The route generator is incorrectly adding a slash to the `$tab/index` route, breaking our app:

<img width="398" alt="Screenshot 2024-04-24 at 2 56 11 PM" src="https://github.com/TanStack/router/assets/5452298/afc52c6f-6b3f-4ba3-b2cb-bbab36864546">

Without this slash the router works as expected but the generator continues to add a slash.

### Your Example Website or App

It's our own internal app.

### Steps to Reproduce the Bug or Issue

Say I have this folder structure:

<img width="202" alt="Screenshot 2024-04-24 at 2 55 31 PM" src="https://github.com/TanStack/router/assets/5452298/549ea738-6b32-40e2-8415-0d47b3d9ac1d">

The route generator is incorrectly adding a slash to the `$tab/index` route, breaking our app:

<img width="398" alt="Screenshot 2024-04-24 at 2 56 11 PM" src="https://github.com/TanStack/router/assets/5452298/afc52c6f-6b3f-4ba3-b2cb-bbab36864546">

Without this slash the router works as expected but the generator continues to add a slash.

### Expected behavior

A slash is not added.

### Platform

macOS
Latest version of TanStack Router 1.30

**Comment by divmgl** [comment]
We were able to fix this by moving `$tab/index.tsx` to `$tab.tsx`. I'm guessing this is intentional?

**Comment by divmgl** [comment]
Just kidding that did not fix the issue. We now have a new issue:

<img width="562" alt="Screenshot 2024-04-24 at 3 07 34 PM" src="https://github.com/TanStack/router/assets/5452298/5c146700-c8c9-4519-a0d5-17ce1e1acd42">

**Comment by divmgl** [comment]
We had to remove TypeScript support for TanStack Router to unblock all of these issues.

* `$tab/index.tsx` creates a trailing slash that breaks our pages
* `$tab.tsx` works but now all of the Links complain
* As of the latest version we have to add `from=""` to links
* Having a parameter in `from=` means that the parameter is expected to be supplied in `params` or it's `undefined`. This is either a regression from previous versions or it's intended behavior that is not clearly documented.

```tsx
// declare module "@tanstack/react-router" {
//  interface Register {
//    router: typeof router
//  }
// }
```

We're also adding to add `from=""` to all of our links to stop TypeScript from complaining. And if we specify the `from` then the params will come up as `undefined`. Older versions of TanStack Router did not require the `from` parameter and the router still worked great supplying incomplete params (which IMHO is the best DX one could ask for).

**Comment by divmgl** [comment]
Also your documentation website is down...

**Comment by andrejilderda** [comment]
I may be overseeing something, but didn't you mean to name your file `route.tsx` in stead of `index.tsx`?

**Comment by schiller-manuel** [comment]
what exactly is "breaking" your app?

please always provide a minimal complete example, e.g. by forking and modifying one of the existing examples on stackblitz

**Comment by divmgl** [comment]
@schiller-manuel https://github.com/divmgl/tanstack-router-bug
<img width="1920" alt="Screenshot 2024-04-24 at 4 57 02 PM" src="https://github.com/TanStack/router/assets/5452298/45048c78-aa31-4022-88c9-e686bcde091c">

**Comment by divmgl** [comment]
<img width="541" alt="Screenshot 2024-04-24 at 4 57 21 PM" src="https://github.com/TanStack/router/assets/5452298/5aad5647-ecec-402e-b3df-13ca9dbb843b">

**Comment by divmgl** [comment]
<img width="1320" alt="Screenshot 2024-04-24 at 4 57 35 PM" src="https://github.com/TanStack/router/assets/5452298/3a65a8ac-4a61-4cf9-8433-6ea4a9aafe32">

**Comment by schiller-manuel** [comment]
so does "breaking" mean typescript issues or runtime behavior?

**Comment by divmgl** [comment]
Changing this to `index.tsx` breaks the links because it adds a trailing slash, just a heads up. So when you click on the links the application goes to a Not Found route.

 That's what I meant by "breaks the app".

**Comment by divmgl** [comment]
Both. TypeScript is not working correctly and a trailing slash will send us to a Not Found page on existing links if we use `index.tsx`.

**Comment by divmgl** [comment]
However the biggest blocker by far is that TypeScript is not working correctly.

**Comment by divmgl** [comment]
Will do. Are you able to see the TS issue though?

**Comment by schiller-manuel** [comment]
we are aware that a recent change that restricts to only link to leaf routes was too rigid and this will be loosened up soon.

however, in your particular case, you might really need both `index.tsx` and `route.tsx`. 
`route.tsx` then contains all the layout that should be displayed for all children routes (e.g. `/projects/123/456/activities/789`), including the index route defined in `index.tsx` (e.g. `/projects/123/456`).

**Comment by divmgl** [comment]
Ah, very interesting. So we needed both an `index.tsx` and `route.tsx` file here. Kind of confusing, should definitely make a note about this in the documentation.

One additional question, is this intentional?

<img width="385" alt="Screenshot 2024-04-24 at 5 21 10 PM" src="https://github.com/TanStack/router/assets/5452298/770fda07-e56d-4105-92dc-8c3156254e06">

Where if `$workspaceId` and `$projectId` are not specified in the params then TypeScript complains? We found it really useful to be able to supply partial params and rely on `@tanstack/react-router`'s functionality where it reuses the existing params using the current route.

**Comment by schiller-manuel** [comment]
For partial params to work, you need to be navigating away `from` a route that has those missing params defined (i.e. the `from` property needs to be set. Otherwise TS does not know in which route the link is created).
However, as you already stated above there currently is a bug (https://github.com/TanStack/router/issues/1474) that prevents the params to be carried over at runtime.

**Comment by schiller-manuel** [comment]
> Ah, very interesting. So we needed both an `index.tsx` and `route.tsx` file here. Kind of confusing, should definitely make a note about this in the documentation.

Did you have a look at https://tanstack.com/router/v1/docs/framework/react/guide/route-trees ?
If this misses some information, we would really appreciate a PR to update the documentation.

**Comment by divmgl** [comment]
> For partial params to work, you need to be navigating away `from` a route that has those missing params defined (i.e. the `from` property needs to be set. Otherwise TS does not know in which route the link is created). However, as you already stated above there currently is a bug (#1474) that prevents the params to be carried over at runtime.

So a previous version of TanStack Router did not work like this and we liked the DX quite a bit. The reason is that links are usually static and are only used in a single place. It'd be nice if this was opt-in or opt-out in some way. Maybe a feature request.

**Comment by divmgl** [comment]
Closing this issue as technically our problem here is resolved.

We're running into other issues with TanStack Router. One additional issue is performance in our TS Server is slowing down considerably as we iterate on routes.

I've opened a discussion here as I technically will not have time to create a repro but I'd be happy to coordinate with you all if you'd like to take a look at our codebase and be compensated for helping us. My email is dexter@permitflow.com.
